keps/sig-architecture/4402-go-workspaces/README.md

```diff
@@ -0,0 +1,374 @@
+off.  Worse, many of our tools make dubious assumptions about the
+interchangeability of Go package paths (e.g. "example.com/foo/bar") and dis
+paths.  Those assumptions held as long as we used GOPATH, BARELY held with
+staging modules, and break totally in workspaces.
+
+This means that some of our ecosystem-facing APIs (codegen tools, in
+particular) *must* change.  We can mitigate some of the impact by creating "v2"
+of things like k8s.io/gengo (which powers many tools).  We might choose to do
+the same in kube-openapi, though its API is not changing, just the
+implementation.
+
+However, k8s.io/code-generator is more challenging.  It holds all of our tools
+and some scripts for invoking them, and is used by a non-trivial number of
+downstream projects (operators, custom API servers, etc).  Making a v2 of this
+(which itself lives in staging, and is subject to this KEP!!) is something we
+have not done before.  The simple answer is to rely on the fact that it was
+never really versioned before, and just make the changes.
+
+Another risk is that this depends on
+[Go 1.22](https://tip.golang.org/doc/go1.22), which is not yet released, and is
+scheduled to release in February 2024.  This seems like good timing for k8s
+1.30, but has not yet been confirmed as plan-of-record.
+
+Another risk is that there are many tools in use which DO NOT HAVE TESTS.  They
+seem to work today, but upon closer inspection it's not always clear that they
+work correctly or completely, or what the exact semantics are intended to be.
+Changing these is dangerous.  As we proceed with this work, we will have to add
+tests to tools and verifiersm, and look closely at the results of
+code-generators.  Ideally there will be NO CHANGE in generated code.
+
+### Test Plan
+
+We will add some tests to untested tools, though complete functional coverage
+is difficult, and would be a prohibitively high bar (e.g. it's unreasonable to
+demand net-new tests for every codegen tool which exercises every path). That
+said, we will look carefully at how each tool works, to mitigate risks.  At the
+time of this writing, the author has spent upwards of 100 hours in the
+debugger, single-stepping through each tool.
+
+[X] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+We will add tests for critical subsystems and tools being changed (e.g. gengo's
+package loading).
+
+##### Integration tests
+
+N/A - see e2e
+
+##### e2e tests
+
+The best test for work like this is, more or less, "did it work?".
+
+If the codegen tools emit the exact same code (or changes are manually
+verifiable as correct) then "it works".  Part of this process must be to
+exercise it on out-of-tree consumers, which will be difficult to do until it is
```

Comment:
it is not the best thing, but it seems we need to choose into keep the current state forever or asking the ecosystem to catchup ... I think is a fair request from Kubernetes

Reply:
Agree, just needed to call it out.